### PR TITLE
More compose

### PR DESCRIPTION
**Pull request by kmcnellis**
### Description
- Allow for docker compose to be used in more commands, removing extra hardcoded arguments.  
- Remove docker-compose container_names, allowing for multiple compose projects to be running at once.
- Simplify the browser test docker-compose file to mostly inherit from the base docker-compose file
- Allow dev, unit test, and browser tests to run at the same time locally
- Rename bin/dump-dev-program-config -> bin/dev-dump-program-config and allow specifying file name for dumping 

### Checklist
- [x] Ran the affected scripts

### CODE REVIEW

File: bin/run-test-ci [hunk 1]
```diff
@@ -3,16 +3,19 @@
 # DOC: Run the unit tests in CI mode.
 
 export COMPOSE_PROJECT_NAME=test-support
-
 source bin/lib.sh
 
-docker::ensure_unit_test_env
```
Comment (shanemc-goog) [inline bin/run-test-ci:9]: What's the general thought behind removing this utility and inline it?

File: civiform.code-workspace [hunk 2]
```diff
@@ -12,5 +12,8 @@
 			"**/target": true
 		},
 		"java.configuration.updateBuildConfiguration": "automatic"
+	},
+	"extensions": {
+		"recommendations": []
```
Comment (shanemc-goog) [inline civiform.code-workspace:17]: what is this file and these lines?  Is it related to compose?

File: docker-compose.dev.yml [hunk 3]
```diff
@@ -33,10 +32,27 @@ services:
...
   db:
     image: postgres:12.5
+    stdin_open: true
+    tty: true
     volumes:
       # isolate and persist dev database.
+      - db-data:/var/lib/postgresql/data
       - project-data:/project
+
+  pgadmin:
```
Comment (shanemc-goog) [inline docker-compose.dev.yml:42]: with the additions here and in the unit test yml files are you removing the composition with docker-compose.yml or is something else going on?

File: browser-test/browser-test-compose.yml [hunk 4]
```diff
@@ -1,67 +1,37 @@
-# Use postgres/example user/passwordcredentials
-version: '3.1'
+# Builds on docker-compose.yml
+version: "3.4"
 
 services:
-
   localstack:
-    image: localstack/localstack
```
Comment (shanemc-goog) [inline browser-test/browser-test-compose.yml:7]: Where are the images being specified now?  Were these redundant with the "base" yml file?

**Review by shanemc-goog — COMMENTED** [review]
As much as I understand docker configs this looks good, had a few questions.

File: bin/attach-dev [hunk 5]
```diff
@@ -20,4 +20,4 @@ echo "Civiform starting locally at http://localhost:9000/loginForm"
 
 echo "Run bin/stop-dev to stop the server"
 
-docker attach civiform
+docker attach "${COMPOSE_PROJECT_NAME}-civiform-1"
```
Comment (goldblatt) [inline bin/attach-dev:23]: i only see this getting set for the tests is it supposed to get set somewhere for just running these commands too?

**Review by goldblatt — CHANGES_REQUESTED** [review]
Looks good just a question around the env variables. Otherwise makes sense to me!

File: bin/dev-pgadmin [hunk 6]
```diff
@@ -24,17 +24,12 @@ tee /tmp/servers.json > /dev/null <<EOT
...
-  -e 'PGADMIN_CONFIG_MASTER_PASSWORD_REQUIRED=False' \
-  -v /tmp/servers.json:/pgadmin4/servers.json \
-  -d dpage/pgadmin4 
+docker compose \
+    -f docker-compose.yml \
+    -f docker-compose.dev.yml \
+    --profile pgadmin \
+    up pgadmin \
+    --wait \
+    -d
```
Comment (bion) [inline bin/dev-pgadmin:33]: this isn't covered by a test, so please confirm the script still works

File: bin/dev-psql [hunk 7]
```diff
@@ -6,10 +6,18 @@ source bin/lib.sh
...
+${COMPOSE_CMD} \
+  up db \
+  --no-deps \
+  --wait \
+  -d
+${COMPOSE_CMD} exec \
   -e "PGPASSWORD=example" \
-  --network "${DOCKER_NETWORK_NAME}" \
-  postgres:12.5 \
+  db \
```
Comment (bion) [inline bin/dev-psql:20]: this isn't covered by a test, so please confirm the script still works with manual testing

File: bin/dev-restore-db [hunk 8]
```diff
@@ -29,21 +29,27 @@ TRUNCATE_SQL="${TRUNCATE_SQL//$'\n'/}"
...
-  -v "$(pwd):/project" \
-  -e "PGPASSWORD=example" \
-  --network "${DOCKER_NETWORK_NAME}" \
-  postgres:12.5 \
-  /bin/bash -c "${CMD}"
+# command for invoking pg_restore with the dev database
+${COMPOSE_CMD} \
+  exec -it -e "PGPASSWORD=example" db \
+  /usr/bin/pg_restore -h db -U postgres \
+   -f "${DUMP_FILE}"
```
Comment (bion) [inline bin/dev-restore-db:55]: this isn't covered by a test, so please confirm the script still works with manual testing

File: bin/dump-dev-program-config [hunk 9]
```diff
@@ -21,4 +16,17 @@ docker run --rm -it \
...
+  -f docker-compose.dev.yml"
+
+${COMPOSE_CMD} \
+  up db \
+  --no-deps \
+  --wait \
+  -d
+
+${COMPOSE_CMD} exec -it -e "PGPASSWORD=example" db ${CMD}
+${COMPOSE_CMD} cp db:/project/dev_programs.dump dev_programs.dump
```
Comment (bion) [inline bin/dump-dev-program-config:32]: this isn't covered by a test, so please confirm the script still works with manual testing

File: bin/attach-dev (hunk 5 above)
Comment (kmcnellis) [reply bin/attach-dev:23]: If you checkout civiform in a different folder, docker uses that folder as the prefix.  COMPOSE_PROJECT_NAME captures that and generates the container name accordingly.

Comment (kmcnellis) [reply bin/attach-dev:23]: It's mostly just unfortunate that there's no docker-compose command for this

File: bin/dev-restore-db (hunk 8 above)
Comment (kmcnellis) [reply bin/dev-restore-db]: confirmed

**Review by bion — DISMISSED** [review]

File: docker-compose.dev.yml [hunk 10]
```diff
@@ -21,23 +21,52 @@ services:
...
-    container_name: civiform-shell
     image: civiform-dev
     entrypoint: /bin/bash
     profiles: ["shell"]
     volumes: *sbt-volumes
     stdin_open: true # docker run -i
     tty: true        # docker run -t
 
+  localstack:
+    ports:
```
Comment (goldblatt) [inline docker-compose.dev.yml:38]: nit: can this be expose instead of the port mapping

**Review by goldblatt — APPROVED** [review]

**Review by shanemc-goog — APPROVED** [review]

**Review by gwendolyngoetz — DISMISSED** [review]

File: docker-compose.dev.yml (hunk 10 above)
Comment (kmcnellis) [reply docker-compose.dev.yml:38]: Unfortunately, no.  We use the ports externally in our startup scripts to poll for the emulators being up.

File: docker-compose.dev.yml (hunk 3 above)
Comment (kmcnellis) [reply docker-compose.dev.yml:53]: I really like it the pgadmin!  Its just in here to make things easier to map together with the db volume.

**Comment by kmcnellis** [comment]
Sorry, need to re-get reviews, since I rebased to master!

File: bin/dev-dump-program-config [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    -h db \
+    -U postgres \
+    -t programs \
+    -t questions \
+    -t versions \
+    -t versions_programs \
+    -t versions_questions \
+    postgres -f /project/${DUMPFILE}"
+
+${COMPOSE_CMD} \
```
Comment (goldblatt) [inline bin/dev-dump-program-config:22]: just a note that this i think is supposed to be $() in a subshell and all in quotes according to the style guide? as far as i can tell but it's not super clear on putting commands in variables

**Review by goldblatt — APPROVED** [review]

**Review by bion — APPROVED** [review]

File: bin/dev-dump-program-config (hunk 11 above)
Comment (kmcnellis) [reply bin/dev-dump-program-config:22]: Not in this case - We're saving a command and some of it's arguments to a variable, so quoting it here will prevent them from splitting into different values and correctly getting run.

**Pull request merged**